File: integrations/rails/app/controllers/todos_controller.rb
# frozen_string_literal: true

# Todo page (with/without @client markers) + the session-cookie-backed,
# Mutex-guarded in-memory REST API. Direct port of the Sinatra example's todo
# routes.
class TodosController < ApplicationController
  def index
    session = bf_session
    todos = session[:todos].map(&:dup)
    done = todos.count { |t| t[:done] }
    component = params[:ssr] ? 'TodoAppSSR' : 'TodoApp'
    render_component(component,
                     children: { 'todo_item' => 'TodoItem' },
                     props: { initialTodos: todos },
                     stash: { todos: todos, newText: '', filter: 'all', doneCount: done })
  end

  # --- todo REST API ---
  def api_index
    render json: bf_session[:todos]
  end

  def api_create
    session = bf_session
    input = parse_json_body
    todo = nil
    # id assignment + increment must be atomic together (Puma is threaded) or
    # two concurrent POSTs could read the same next_id before either increments.
    Barefoot::SESSIONS_MUTEX.synchronize do
      todo = { id: session[:next_id], text: input[:text], done: false, editing: false }
      session[:todos].push(todo)
      session[:next_id] += 1
    end
    render json: todo, status: :created
  end

  def api_update
    session = bf_session
    input = parse_json_body
    id = params[:id].to_i
    todo = Barefoot::SESSIONS_MUTEX.synchronize do
      t = session[:todos].find { |x| x[:id] == id }
      next nil unless t

      t[:text] = input[:text] if input.key?(:text)
      t[:done] = !!input[:done] if input.key?(:done)
      t
    end
    return render json: { error: 'not found' }, status: :not_found unless todo

    render json: todo
  end

  def api_destroy
    session = bf_session
    id = params[:id].to_i
    Barefoot::SESSIONS_MUTEX.synchronize { session[:todos].reject! { |t| t[:id] == id } }
    head :no_content
  end

  def api_reset
    session = bf_session
    Barefoot::SESSIONS_MUTEX.synchronize do
      session[:todos] = Barefoot.seed_todos
      session[:next_id] = 4
    end
    render plain: 'ok'
  end
end

File: integrations/laravel/app/Http/Controllers/TodosController.php
<?php

declare(strict_types=1);

namespace App\Http\Controllers;

use App\Support\ExampleApp;
use App\Support\TodoStore;
use Illuminate\Http\JsonResponse;
use Illuminate\Http\Request;
use Illuminate\Http\Response;
use Symfony\Component\HttpFoundation\Response as BaseResponse;

/**
 * Todo pages (with/without @client markers) + the session-cookie REST API.
 * Port of integrations/blade's todo routes; the store is the same
 * file-backed, flock-guarded JSON store (see App\Support\TodoStore for why
 * an in-memory array doesn't survive `PHP_CLI_SERVER_WORKERS=8`).
 */
final class TodosController extends Controller
{
    public function index(Request $request, ?string $ssr = null): Response
    {
        [$sid, $minted] = $this->resolveSessionId($request);
        $state = TodoStore::read($sid);
        $todos = $state['todos'];
        $done = count(array_filter($todos, static fn ($t) => $t['done']));
        $component = $ssr !== null ? 'TodoAppSSR' : 'TodoApp';
        $response = response($this->renderComponent(
            $component,
            children: ['todo_item' => 'TodoItem'],
            props: ['initialTodos' => $todos],
            stash: ['todos' => $todos, 'newText' => '', 'filter' => 'all', 'doneCount' => $done],
        ));
        return $this->withSessionCookie($response, $sid, $minted);
    }

    // --- todo REST API ---

    public function apiIndex(Request $request): BaseResponse
    {
        [$sid, $minted] = $this->resolveSessionId($request);
        $state = TodoStore::read($sid);
        return $this->withSessionCookie(response()->json($state['todos']), $sid, $minted);
    }

    public function apiCreate(Request $request): BaseResponse
    {
        [$sid, $minted] = $this->resolveSessionId($request);
        $body = $request->json()->all();
        $todo = TodoStore::with($sid, static function (array $state) use ($body) {
            $newTodo = ['id' => $state['next_id'], 'text' => $body['text'] ?? null, 'done' => false, 'editing' => false];
            $state['todos'][] = $newTodo;
            $state['next_id']++;
            return [$state, $newTodo];
        });
        return $this->withSessionCookie(response()->json($todo, 201), $sid, $minted);
    }

    public function apiUpdate(Request $request, int $id): BaseResponse
    {
        [$sid, $minted] = $this->resolveSessionId($request);
        $body = $request->json()->all();
        $todo = TodoStore::with($sid, static function (array $state) use ($id, $body) {
            foreach ($state['todos'] as &$t) {
                if ($t['id'] !== $id) {
                    continue;
                }
                if (array_key_exists('text', $body)) {
                    $t['text'] = $body['text'];
                }
                if (array_key_exists('done', $body)) {
                    $t['done'] = (bool) $body['done'];
                }
                return [$state, $t];
            }
            return [$state, null];
        });
        $response = $todo === null
            ? response()->json(['error' => 'not found'], 404)
            : response()->json($todo);
        return $this->withSessionCookie($response, $sid, $minted);
    }

    public function apiDestroy(Request $request, int $id): BaseResponse
    {
        [$sid] = $this->resolveSessionId($request);
        TodoStore::with($sid, static function (array $state) use ($id) {
            $state['todos'] = array_values(array_filter($state['todos'], static fn ($t) => $t['id'] !== $id));
            return [$state, null];
        });
        return response()->noContent();
    }

    public function apiReset(Request $request): BaseResponse
    {
        [$sid, $minted] = $this->resolveSessionId($request);
        TodoStore::with($sid, static fn (array $state) => [['todos' => ExampleApp::seedTodos(), 'next_id' => 4], null]);
        return $this->withSessionCookie(response('ok')->header('Content-Type', 'text/plain'), $sid, $minted);
    }

    /** Attach the raw `bf_session` cookie only when a new id was minted
     * (mirrors integrations/blade's `if ($minted) set_session_cookie(...)`
     * call sites). */
    private function withSessionCookie(BaseResponse $response, string $sid, bool $minted): BaseResponse
    {
        if ($minted) {
            $response->headers->setCookie($this->sessionCookie($sid));
        }
        return $response;
    }
}
